subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/DefaultWatchingVirtualFileSystem.java

```diff
@@ -68,64 +68,58 @@ public void startWatching() {
                     return;
                 }
 
-                try {
-                    // For existing files and directories we watch the parent directory,
-                    // so we learn if the entry itself disappears or gets modified.
-                    // In case of a missing file we need to find the closest existing
-                    // ancestor to watch so we can learn if the missing file respawns.
-                    Path ancestor = path;
-                    while (true) {
-                        ancestor = ancestor.getParent();
-                        if (ancestor == null) {
-                            break;
-                        }
-                        if (Files.exists(ancestor)) {
-                            watch(ancestor, visited);
-                            break;
-                        }
+                // For existing files and directories we watch the parent directory,
+                // so we learn if the entry itself disappears or gets modified.
+                // In case of a missing file we need to find the closest existing
+                // ancestor to watch so we can learn if the missing file respawns.
+                Path ancestor = path;
+                while (true) {
+                    ancestor = ancestor.getParent();
```

Comment:
When adding a parent we should make a note that we only care about the one location in the parent and nothing else. So we could filter the events in the OS-specific watchers better.

That is for later I suppose.

Reply:
Yes, this is pretty rough at present. IIRC, OSs exhibit both behaviors when watching a directory: a) report changes to descendants only, b) report changes to descendants and self.

For those that do b) we could optimize this even more. I'd keep this PR as is for now, though.